pkg/logentry/stages/timestamp.go

```diff
@@ -67,49 +94,130 @@ func newTimestampStage(logger log.Logger, config interface{}) (*timestampStage,
-			}
+
+	parsedTs, err := ts.parseTimestampFromSource(extracted)
+	if err != nil {
+		ts.processActionOnFailure(labels, t)
+		return
+	}
+
+	// Update the log entry timestamp with the parsed one
+	*t = *parsedTs
+
+	// The timestamp has been correctly parsed, so we should store it in the map
+	// containing the last known timestamp used by the "fudge" action on failure.
+	if *ts.cfg.ActionOnFailure == TimestampActionOnFailureFudge {
+		ts.lastKnownTimestamps.Add(ts.getStreamID(labels), *t)
+	}
+}
+
+func (ts *timestampStage) parseTimestampFromSource(extracted map[string]interface{}) (*time.Time, error) {
+	// Ensure the extracted data contains the timestamp source
+	v, ok := extracted[ts.cfg.Source]
+	if !ok {
+		if Debug {
+			level.Debug(ts.logger).Log("msg", ErrTimestampSourceMissing)
 		}
 
-		parsedTs, err := ts.parser(s)
-		if err != nil {
-			if Debug {
-				level.Debug(ts.logger).Log("msg", "failed to parse time", "err", err, "format", ts.cfgs.Format, "value", s)
-			}
-		} else {
-			*t = parsedTs
+		return nil, errors.New(ErrTimestampSourceMissing)
+	}
+
+	// Convert the timestamp source to string (if it's not a string yet)
+	s, err := getString(v)
+	if err != nil {
+		if Debug {
+			level.Debug(ts.logger).Log("msg", ErrTimestampConversionFailed, "err", err, "type", reflect.TypeOf(v).String())
 		}
-	} else {
+
+		return nil, errors.New(ErrTimestampConversionFailed)
+	}
+
+	// Parse the timestamp source according to the configured format
+	parsedTs, err := ts.parser(s)
+	if err != nil {
 		if Debug {
-			level.Debug(ts.logger).Log("msg", "extracted data did not contain a timestamp")
+			level.Debug(ts.logger).Log("msg", ErrTimestampParsingFailed, "err", err, "format", ts.cfg.Format, "value", s)
 		}
+
+		return nil, errors.New(ErrTimestampParsingFailed)
 	}
+
+	return &parsedTs, nil
 }
 
-// Name implements Stage
-func (ts *timestampStage) Name() string {
-	return StageTypeTimestamp
+func (ts *timestampStage) processActionOnFailure(labels model.LabelSet, t *time.Time) {
+	switch *ts.cfg.ActionOnFailure {
+	case TimestampActionOnFailureFudge:
+		ts.processActionOnFailureFudge(labels, t)
+	case TimestampActionOnFailureSkip:
+		// Nothing to do
+	}
+}
+
+func (ts *timestampStage) processActionOnFailureFudge(labels model.LabelSet, t *time.Time) {
+	streamID := ts.getStreamID(labels)
+	lastTimestamp, ok := ts.lastKnownTimestamps.Get(streamID)
+
+	// If the last known timestamp is unknown (ie. has not been successfully parsed yet)
+	// there's nothing we can do, so we're going to keep the current timestamp
+	if !ok {
+		return
+	}
+
+	// Fudge the timestamp
+	*t = lastTimestamp.(time.Time).Add(1 * time.Nanosecond)
+
+	// Store the fudged timestamp, so that a subsequent fudged timestamp will be 1ns after it
+	ts.lastKnownTimestamps.Add(streamID, *t)
+}
+
+func (ts *timestampStage) getStreamID(labels model.LabelSet) string {
+	// If the filename label is set, we assume it's a file target
+	if value, ok := labels["filename"]; ok {
```

Comment:
I think you should use `FastFingerprint()` from the `model.LabelSet` and use that for your map.